File: analysis/import/CSVImporter/src/main/java/de/maibornwolff/codecharta/importer/csv/CSVRow.java
package de.maibornwolff.codecharta.importer.csv;

import java.util.Arrays;
import java.util.Map;
import java.util.function.Function;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class CSVRow {
    private static final Pattern FLOAT_PATTERN = Pattern.compile("\\d+[,.]?\\d*");

    private final String[] row;
    private final CSVHeader header;
    private final char pathSeparator;

    public CSVRow(String[] row, CSVHeader header, char pathSeparator) {
        if (row.length <= header.getPathColumn()) {
            throw new IllegalArgumentException(
                    "Row " + Arrays.toString(row) + " has no column containing the file path. Should be in " + header.getPathColumn() + "th column.");
        }

        this.pathSeparator = pathSeparator;
        this.row = row;
        this.header = header;
    }

    public String getPath() {
        return row[header.getPathColumn()];
    }

    public String getFileName() {
        String path = getPath();
        return path.substring(path.lastIndexOf(pathSeparator) + 1);
    }

    public String getFolderWithFile() {
        String path = getPath();
        return path.substring(0, path.lastIndexOf(pathSeparator) + 1);
    }

    private boolean validAttributeOfRow(int i) {
        return i < row.length && row[i] != null && FLOAT_PATTERN.matcher(row[i]).matches();
    }

    public Map<String, Object> getAttributes() {
        Function<Integer, String> attributeNameLambda = i -> header.getColumnName(i);
        Function<Integer, Object> attributeValueLambda = i -> Float.parseFloat(row[i].replace(',', '.'));

        return header.getColumnNumbers()
                .stream()
                .filter(this::validAttributeOfRow)
                .collect(
                        Collectors.toMap(
                                attributeNameLambda,
                                attributeValueLambda
                        )
                );
    }
}

File: analysis/import/CSVImporter/src/main/kotlin/de/maibornwolff/codecharta/importer/csv/CSVRow.kt
package de.maibornwolff.codecharta.importer.csv

import de.maibornwolff.codecharta.model.MutableNode
import de.maibornwolff.codecharta.model.NodeType
import de.maibornwolff.codecharta.model.Path
import de.maibornwolff.codecharta.model.PathFactory
import java.util.regex.Pattern

class CSVRow(private val row: Array<String?>, private val header: CSVHeader, private val pathSeparator: Char) {

    init {
        if (row.size <= header.pathColumn) {
            throw IllegalArgumentException(
                "Row  has no column containing the file path. Should be in ${header.pathColumn} th column."
            )
        }
    }

    fun pathInTree(): Path {
        return PathFactory.fromFileSystemPath(
            path.substring(0, path.lastIndexOf(pathSeparator) + 1),
            pathSeparator
        )
    }

    fun asNode(): MutableNode {
        val filename = path.substring(path.lastIndexOf(pathSeparator) + 1)
        return MutableNode(filename, NodeType.File, attributes)
    }

    private val path =
        if (row[header.pathColumn] == null) throw IllegalArgumentException("Row has no path information.")
        else row[header.pathColumn]!!
    private val floatPattern = Pattern.compile("\\d+[,.]?\\d*")
    private fun validAttributeOfRow(i: Int) =
        i < row.size && row[i] != null && floatPattern.matcher(row[i]).matches()

    private fun parseAttributeOfRow(i: Int) =
        java.lang.Double.parseDouble(row[i]!!.replace(',', '.'))

    private val attributes =
        header.columnNumbers
            .filter { validAttributeOfRow(it) }
            .associateBy(
                { header.getColumnName(it) },
                { parseAttributeOfRow(it) }
            )
}
